File: src/mtg_deck_tools/builder/deck_metrics.py
"""Post-build deck composition metrics (UX10a)."""

from __future__ import annotations

from mtg_deck_tools.builder.curve_advisories import (
    curve_advisory_blurb,
    evaluate_curve_advisories,
)
from mtg_deck_tools.builder.deck import DeckCard
from mtg_deck_tools.builder.mana_base import ManaBasePlan

_CMC_BUCKETS = tuple(str(n) for n in range(8)) + ("7+",)
_TYPE_ORDER = (
    "Creature",
    "Vehicle",
    "Instant",
    "Sorcery",
    "Artifact",
    "Enchantment",
    "Equipment",
    "Planeswalker",
    "Battle",
    "Land",
    "Other",
)

_FILTER_SUBTYPES = ("Vehicle", "Equipment")


def _is_land(type_line: str) -> bool:
    return "Land" in (type_line or "")


def _is_creature(type_line: str) -> bool:
    line = type_line or ""
    return "Creature" in line and "Vehicle" not in line


def primary_card_type(type_line: str) -> str:
    """Last recognized primary type before the em dash, matching web deck-view."""
    normalized = (type_line or "").strip()
    for subtype in _FILTER_SUBTYPES:
        if subtype in normalized:
            return subtype
    before_dash = normalized.split("—")[0].strip()
    parts = before_dash.split()
    for part in reversed(parts):
        if part in _TYPE_ORDER:
            return part
    return parts[-1] if parts else "Other"


def _cmc_bucket(cmc: float) -> str:
    if cmc >= 7:
        return "7+"
    return str(int(cmc))


def cmc_histogram(
    cards: list[DeckCard],
    *,
    creatures_only: bool = False,
    exclude_lands: bool = True,
) -> dict[str, int]:
    """Quantity-weighted CMC histogram with buckets 0–7 and 7+."""
    counts = {bucket: 0 for bucket in _CMC_BUCKETS}
    for card in cards:
        type_line = card.type_line or ""
        if exclude_lands and _is_land(type_line):
            continue
        if creatures_only and not _is_creature(type_line):
            continue
        bucket = _cmc_bucket(card.cmc)
        counts[bucket] += card.quantity
    return counts


def type_counts(cards: list[DeckCard]) -> dict[str, int]:
    """Quantity-weighted counts by primary card type."""
    counts: dict[str, int] = {}
    for card in cards:
        key = primary_card_type(card.type_line or "")
        counts[key] = counts.get(key, 0) + card.quantity
    return counts


def _avg_cmc(cards: list[DeckCard], *, predicate) -> float | None:
    total_cmc = 0.0
    total_qty = 0
    for card in cards:
        if not predicate(card.type_line or ""):
            continue
        total_cmc += card.cmc * card.quantity
        total_qty += card.quantity
    if total_qty == 0:
        return None
    return round(total_cmc / total_qty, 2)


def compute_deck_metrics(
    cards: list[DeckCard],
    *,
    mana_base: ManaBasePlan | None = None,
) -> dict:
    """Build UX10a metrics from maindeck card rows."""
    nonland_predicate = lambda tl: not _is_land(tl)  # noqa: E731
    creature_predicate = lambda tl: _is_creature(tl)  # noqa: E731

    land_count = sum(card.quantity for card in cards if _is_land(card.type_line or ""))
    ramp_count = mana_base.ramp.total if mana_base else sum(
        card.quantity for card in cards if card.slot == "ramp"
    )

    return {
        "cmc_histogram": cmc_histogram(cards),
        "creature_cmc_histogram": cmc_histogram(cards, creatures_only=True),
        "type_counts": type_counts(cards),
        "avg_cmc_nonland": _avg_cmc(cards, predicate=nonland_predicate),
        "avg_creature_cmc": _avg_cmc(cards, predicate=creature_predicate),
        "land_count": land_count,
        "ramp_count": ramp_count,
    }


_BAR_WIDTH = 20


def _ascii_bar(count: int, max_count: int) -> str:
    if count <= 0 or max_count <= 0:
        return ""
    filled = max(1, round(count / max_count * _BAR_WIDTH))
    return "█" * filled


def _curve_blurb(
    cmc_histogram: dict[str, int],
    *,
    themes: list[str] | None = None,
    metrics: dict | None = None,
) -> str:
    total = sum(cmc_histogram.values())
    if total == 0:
        return "No nonland spells to chart."
    base_metrics = metrics or {"cmc_histogram": cmc_histogram}
    advisories = evaluate_curve_advisories(base_metrics, themes=themes)
    return curve_advisory_blurb(advisories, histogram="nonlands")


def render_deck_metrics_section(
    metrics: dict,
    *,
    themes: list[str] | None = None,
) -> list[str]:
    """Render ## Deck metrics markdown lines."""
    lines = ["## Deck metrics", ""]
    counts_by_type: dict[str, int] = metrics.get("type_counts") or {}
    if counts_by_type:
        lines.extend(["### Card types", "", "| Type | Count |", "| --- | ---: |"])
        for key in sorted(counts_by_type, key=lambda k: (-counts_by_type[k], k)):
            lines.append(f"| {key} | {counts_by_type[key]} |")
        lines.append("")

    cmc_hist: dict[str, int] = metrics.get("cmc_histogram") or {}
    creature_hist: dict[str, int] = metrics.get("creature_cmc_histogram") or {}
    curve_advisories = metrics.get("curve_advisories")
    if curve_advisories is None:
        curve_advisories = [
            item.to_dict()
            for item in evaluate_curve_advisories(metrics, themes=themes)
        ]
    if cmc_hist:
        max_count = max(cmc_hist.values(), default=0)
        lines.extend(
            [
                "### Mana curve (nonlands)",
                "",
                f"_{_curve_blurb(cmc_hist, themes=themes, metrics=metrics)}_",
                "",
                "| CMC | Count | Chart |",
                "| --- | ---: | --- |",
            ]
        )
        for bucket in _CMC_BUCKETS:
            count = cmc_hist.get(bucket, 0)
            lines.append(f"| {bucket} | {count} | {_ascii_bar(count, max_count)} |")
        lines.append("")

    if curve_advisories:
        lines.extend(["### Curve advisories", ""])
        for item in curve_advisories:
            rule = item.get("rule", "CURVE")
            message = item.get("message", "")
            lines.append(f"- **[{rule}]** {message}")
        lines.append("")

    avg_nonland = metrics.get("avg_cmc_nonland")
    avg_creature = metrics.get("avg_creature_cmc")
    land_count = metrics.get("land_count")
    ramp_count = metrics.get("ramp_count")
    summary_parts: list[str] = []
    if avg_nonland is not None:
        summary_parts.append(f"avg nonland CMC **{avg_nonland}**")
    if avg_creature is not None:
        summary_parts.append(f"avg creature CMC **{avg_creature}**")
    if land_count is not None:
        summary_parts.append(f"**{land_count}** lands")
    if ramp_count is not None:
        summary_parts.append(f"**{ramp_count}** ramp")
    if summary_parts:
        lines.append("### Summary")
        lines.append("")
        lines.append(" · ".join(summary_parts))
        lines.append("")

    if creature_hist and any(creature_hist.values()):
        max_creature = max(creature_hist.values())
        lines.extend(
            [
                "### Creature curve",
                "",
                "| CMC | Count | Chart |",
                "| --- | ---: | --- |",
            ]
        )
        for bucket in _CMC_BUCKETS:
            count = creature_hist.get(bucket, 0)
            if count == 0:
                continue
            lines.append(f"| {bucket} | {count} | {_ascii_bar(count, max_creature)} |")
        lines.append("")

    return lines

File: packages/web/src/lib/deck-metrics.ts
import type { DeckCardRow } from "./deck-view";
import { formatTagLabel } from "./format";

export const CMC_BUCKETS = ["0", "1", "2", "3", "4", "5", "6", "7", "7+"] as const;

export type CmcBucket = (typeof CMC_BUCKETS)[number];
export type CurveView = "nonlands" | "creatures";

export interface CmcHistogram {
  [bucket: string]: number;
}

export interface CurveAdvisory {
  rule: string;
  status: string;
  message: string;
  actual_share: number;
  threshold: number;
  histogram: string;
}

export interface DeckMetrics {
  cmc_histogram: CmcHistogram;
  creature_cmc_histogram: CmcHistogram;
  type_counts: Record<string, number>;
  avg_cmc_nonland: number | null;
  avg_creature_cmc: number | null;
  land_count: number;
  ramp_count: number;
  curve_advisories?: CurveAdvisory[];
}

function asRecord(value: unknown): Record<string, unknown> | null {
  return value && typeof value === "object" ? (value as Record<string, unknown>) : null;
}

function asNumber(value: unknown): number | null {
  if (typeof value === "number" && Number.isFinite(value)) return value;
  return null;
}

function emptyHistogram(): CmcHistogram {
  return Object.fromEntries(CMC_BUCKETS.map((bucket) => [bucket, 0]));
}

function isLand(typeLine: string): boolean {
  return typeLine.toLowerCase().includes("land");
}

function isCreature(typeLine: string): boolean {
  return typeLine.includes("Creature") && !typeLine.includes("Vehicle");
}

function cmcBucket(cmc: number): CmcBucket {
  if (cmc >= 7) return "7+";
  return String(Math.trunc(cmc)) as CmcBucket;
}

export function cmcHistogramFromCards(
  cards: DeckCardRow[],
  options: { creaturesOnly?: boolean; excludeLands?: boolean } = {},
): CmcHistogram {
  const { creaturesOnly = false, excludeLands = true } = options;
  const counts = emptyHistogram();
  for (const card of cards) {
    const typeLine = card.type_line ?? "";
    if (excludeLands && isLand(typeLine)) continue;
    if (creaturesOnly && !isCreature(typeLine)) continue;
    const bucket = cmcBucket(card.cmc);
    counts[bucket] = (counts[bucket] ?? 0) + card.quantity;
  }
  return counts;
}

function avgCmc(cards: DeckCardRow[], predicate: (typeLine: string) => boolean): number | null {
  let totalCmc = 0;
  let totalQty = 0;
  for (const card of cards) {
    if (!predicate(card.type_line ?? "")) continue;
    totalCmc += card.cmc * card.quantity;
    totalQty += card.quantity;
  }
  if (totalQty === 0) return null;
  return Math.round((totalCmc / totalQty) * 100) / 100;
}

export function computeDeckMetricsFromCards(cards: DeckCardRow[]): DeckMetrics {
  const typeCounts: Record<string, number> = {};
  let landCount = 0;
  let rampCount = 0;

  for (const card of cards) {
    typeCounts[card.primary_type] = (typeCounts[card.primary_type] ?? 0) + card.quantity;
    if (isLand(card.type_line ?? "")) landCount += card.quantity;
    if (card.slot === "ramp") rampCount += card.quantity;
  }

  return {
    cmc_histogram: cmcHistogramFromCards(cards),
    creature_cmc_histogram: cmcHistogramFromCards(cards, { creaturesOnly: true }),
    type_counts: typeCounts,
    avg_cmc_nonland: avgCmc(cards, (tl) => !isLand(tl)),
    avg_creature_cmc: avgCmc(cards, (tl) => isCreature(tl)),
    land_count: landCount,
    ramp_count: rampCount,
  };
}

function parseHistogram(value: unknown): CmcHistogram | null {
  const row = asRecord(value);
  if (!row) return null;
  const hist = emptyHistogram();
  let hasValue = false;
  for (const bucket of CMC_BUCKETS) {
    const count = asNumber(row[bucket]);
    if (count != null && count > 0) {
      hist[bucket] = Math.max(0, Math.trunc(count));
      hasValue = true;
    }
  }
  return hasValue ? hist : null;
}

function parseTypeCounts(value: unknown): Record<string, number> | null {
  const row = asRecord(value);
  if (!row) return null;
  const counts: Record<string, number> = {};
  for (const [key, raw] of Object.entries(row)) {
    const count = asNumber(raw);
    if (count != null && count > 0) counts[key] = Math.trunc(count);
  }
  return Object.keys(counts).length ? counts : null;
}

function parseCurveAdvisories(value: unknown): CurveAdvisory[] | undefined {
  if (!Array.isArray(value)) return undefined;
  const advisories: CurveAdvisory[] = [];
  for (const item of value) {
    const row = asRecord(item);
    if (!row) continue;
    const rule = typeof row.rule === "string" ? row.rule : "";
    const message = typeof row.message === "string" ? row.message : "";
    const status = typeof row.status === "string" ? row.status : "warn";
    const histogram = typeof row.histogram === "string" ? row.histogram : "nonlands";
    const actualShare = asNumber(row.actual_share);
    const threshold = asNumber(row.threshold);
    if (!rule || !message || actualShare == null || threshold == null) continue;
    advisories.push({
      rule,
      status,
      message,
      actual_share: actualShare,
      threshold,
      histogram,
    });
  }
  return advisories.length ? advisories : undefined;
}

function statsHaveMetrics(stats: Record<string, unknown>): boolean {
  return parseHistogram(stats.cmc_histogram) != null;
}

export function parseDeckMetrics(
  stats: unknown,
  cards: DeckCardRow[],
): DeckMetrics | null {
  const fromCards = computeDeckMetricsFromCards(cards);
  const row = asRecord(stats);
  if (!row || !statsHaveMetrics(row)) {
    const total = sumHistogram(fromCards.cmc_histogram);
    return total > 0 ? fromCards : null;
  }

  return {
    cmc_histogram: parseHistogram(row.cmc_histogram) ?? fromCards.cmc_histogram,
    creature_cmc_histogram:
      parseHistogram(row.creature_cmc_histogram) ?? fromCards.creature_cmc_histogram,
    type_counts: parseTypeCounts(row.type_counts) ?? fromCards.type_counts,
    avg_cmc_nonland: asNumber(row.avg_cmc_nonland) ?? fromCards.avg_cmc_nonland,
    avg_creature_cmc: asNumber(row.avg_creature_cmc) ?? fromCards.avg_creature_cmc,
    land_count: Math.max(0, asNumber(row.land_count) ?? fromCards.land_count),
    ramp_count: Math.max(0, asNumber(row.ramp_count) ?? fromCards.ramp_count),
    curve_advisories: parseCurveAdvisories(row.curve_advisories),
  };
}

export function sumHistogram(histogram: CmcHistogram): number {
  return CMC_BUCKETS.reduce((sum, bucket) => sum + (histogram[bucket] ?? 0), 0);
}

export function maxHistogramCount(histogram: CmcHistogram): number {
  return Math.max(0, ...CMC_BUCKETS.map((bucket) => histogram[bucket] ?? 0));
}

export function histogramForView(metrics: DeckMetrics, view: CurveView): CmcHistogram {
  return view === "creatures" ? metrics.creature_cmc_histogram : metrics.cmc_histogram;
}

export function curveBlurb(histogram: CmcHistogram, advisories?: CurveAdvisory[]): string {
  const total = sumHistogram(histogram);
  if (total === 0) return "No cards to chart for this view.";
  const matching = (advisories ?? []).filter((item) => item.histogram === "nonlands");
  if (matching.length) return matching.map((item) => item.message).join(" ");
  const early =
    (histogram["0"] ?? 0) + (histogram["1"] ?? 0) + (histogram["2"] ?? 0);
  const top =
    (histogram["5"] ?? 0) +
    (histogram["6"] ?? 0) +
    (histogram["7"] ?? 0) +
    (histogram["7+"] ?? 0);
  const earlyShare = early / total;
  const topShare = top / total;
  if (earlyShare < 0.15) return "Light early game — few cards at 0–2 CMC.";
  if (topShare > 0.45) return "Top-heavy curve — many cards at 5+ CMC.";
  return "Mana curve is spread across several CMC bands.";
}

/** Human-readable title for a curve advisory rule id (e.g. CURVE_TOP_HEAVY). */
export function formatCurveAdvisoryTitle(rule: string): string {
  const withoutPrefix = rule.startsWith("CURVE_") ? rule.slice("CURVE_".length) : rule;
  return formatTagLabel(withoutPrefix);
}

export function formatMetricsSummary(metrics: DeckMetrics): string {
  const parts: string[] = [];
  if (metrics.avg_cmc_nonland != null) {
    parts.push(`Avg nonland CMC ${metrics.avg_cmc_nonland.toFixed(1)}`);
  }
  if (metrics.avg_creature_cmc != null) {
    parts.push(`Avg creature CMC ${metrics.avg_creature_cmc.toFixed(1)}`);
  }
  if (metrics.land_count > 0) parts.push(`${metrics.land_count} lands`);
  if (metrics.ramp_count > 0) parts.push(`${metrics.ramp_count} ramp`);
  return parts.join(" · ");
}
